Keep random race (#265)

vasilenkoalexey opened the pull request:
Fix #265

ihhub commented:
Hi @vasilenkoalexey , this solution is not fully correct. For example, I chose a scenario with these settings:
![image](https://user-images.githubusercontent.com/19829520/99551045-5d5cf980-29f6-11eb-94cf-be9c9a221112.png)

Then I restart the scenario and as you could notice my previous choice was not respected:
![image](https://user-images.githubusercontent.com/19829520/99551111-6fd73300-29f6-11eb-8605-5dc64745c512.png)

The issue here is to remember the selection of a race for the last map. If we restart a scenario we should restore these settings. What we need to do is to have for example an array where we store race selection of scenario and apply these selections while we reopen the dialog.

ihhub reviewed (changes requested):
Hi @vasilenkoalexey I'm marking this pull request as changes requested because of the case which I described previously. Please let me know if you need some help in explanation or development.

vasilenkoalexey commented:
I hope I got the change request right

ihhub commented on src/fheroes2/game/game_scenarioinfo.cpp:
```diff
@@ -152,18 +152,21 @@ int Game::ScenarioInfo( void )
         for ( MapsFileInfoList::const_iterator mapIter = lists.begin(); mapIter != lists.end(); ++mapIter ) {
             if ( ( mapIter->name == mapName ) && ( System::GetBasename( mapIter->file ) == mapFileName ) ) {
-                if ( mapIter->file != conf.CurrentFileInfo().file )
+                if ( mapIter->file == conf.CurrentFileInfo().file ) {
```
Here we are changing the original logic to the opposite. Could you please explain the reason behind this?

vasilenkoalexey replied:
Here we try to find map file with same map name and file name as at `conf.CurrentFileInfo()` and if we find successfully we should break search loop and set players info from file which we found. Else - set players info from first file in map list.
I think the original code had a logical error or was not fully completed.

ihhub reviewed (commented):
Hi @vasilenkoalexey , good job! Could you please take a look into given comments?

vasilenkoalexey commented on src/fheroes2/game/game.cpp:
```diff
@@ -69,6 +69,7 @@ namespace Game
     std::string last_name;
     int save_version = CURRENT_FORMAT_VERSION;
     std::vector<int> reserved_vols( LOOPXX_COUNT, 0 );
+    std::map<std::string, StreamBuf> map_players;
```
>Can we just use Players instead of StreamBuf?

Just Players not enough. Look at players.cpp

```
StreamBase & operator<<( StreamBase & msg, const Players & players )
{
    msg << players.GetColors() << players.current_color;

    for ( Players::const_iterator it = players.begin(); it != players.end(); ++it )
        msg << ( **it );

    return msg;
}
```

So we need tuple or custom struct for colors and current color and Players. I tried to reuse the existing code, and I don’t think we will get any significant performance improvement if we use something else instead of StreamBuf, since the maximum number of players is only 6. But if you insist, I can make the necessary changes.

vasilenkoalexey replied:
Hmm... colors and current color need only for game save. Ok, we can use Players here.

vasilenkoalexey commented on src/fheroes2/game/game.cpp:
```diff
@@ -100,6 +101,18 @@ namespace Game
+
+void Game::SavePlayers( const std::string & mapFileName, const Players & players )
+{
+    map_players.insert( std::make_pair( mapFileName, StreamBuf() ) ).first->second << players;
```
> Please use emplace method instead to avoid object copying.

Could you please clarify what you mean?

ihhub commented:
Hi @vasilenkoalexey , I'm going to review all pull request very soon.

ihhub commented on src/fheroes2/game/game.cpp:
```diff
@@ -100,6 +101,40 @@ namespace Game
+{
+    auto it = mapPlayers.lower_bound( mapFileName );
+    if ( it == mapPlayers.end() || mapPlayers.key_comp()( mapFileName, it->first ) ) {
+        it = mapPlayers.insert( it, std::map<std::string, std::vector<Player> >::value_type( mapFileName, std::vector<Player>() ) );
```
Please try this:
```
it = mapPlayers.insert( it, std::make_pair( mapFileName, std::vector<Player>() ) );
```

ihhub replied:
Also there is no need to even specify an iterator for insert function as map sorts everything within this operation. You could just do like this:
```
it = mapPlayers( std::make_pair( mapFileName, std::vector<Player>() ) );
```
```

ihhub commented on src/fheroes2/game/game.cpp:
```diff
@@ -100,6 +101,40 @@ namespace Game
+void Game::SavePlayers( const std::string & mapFileName, const Players & players )
+{
+    auto it = mapPlayers.lower_bound( mapFileName );
+    if ( it == mapPlayers.end() || mapPlayers.key_comp()( mapFileName, it->first ) ) {
```
Could you please explain the use of the second condition?

vasilenkoalexey replied:
So, here i just use most effective way (avoid call `std::vector` constructor if record already exists) for map inserting from "Effective STL", example on stackoverflow [std::map insert or std::map find?](https://stackoverflow.com/questions/97050/stdmap-insert-or-stdmap-find/101980#101980)

ihhub commented on src/fheroes2/game/game.cpp:
```diff
@@ -69,6 +69,7 @@ namespace Game
     std::string last_name;
     int save_version = CURRENT_FORMAT_VERSION;
     std::vector<int> reserved_vols( LOOPXX_COUNT, 0 );
+    std::map<std::string, std::vector<Player>> mapPlayers;
```
I don't think that we even need to store a map. We just need to remember the last loaded map. The same logic exists in the original game. I hope it'll simplify a lot of your changes.

vasilenkoalexey replied:
Ok, than i will remove map and map inserting stuff

ihhub commented on src/fheroes2/system/players.h:
```diff
@@ -174,6 +174,9 @@ class Players : public std::vector<Player *>
 
     int current_color;
+
+    static Player * _players[KINGDOMMAX + 1];
+    static int human_colors;
```
We shouldn't do like this. First of all, all members of a class must be protected or private (ignore the original code). Secondly, static variables must be `const`. A proper way to deal with it is to add a static method to change these variables.

vasilenkoalexey replied:
Done, but... Can't set static vars as `const` because than can't change it... I can move it back to `namespace`...

ihhub reviewed (changes requested):
Hi @vasilenkoalexey , I proposed a simplified way of implementation and also commented some other places. Could you please take a look?

vasilenkoalexey commented on src/fheroes2/game/game.cpp:
```diff
@@ -100,6 +101,40 @@ namespace Game
+void Game::LoadPlayers( const std::string & mapFileName, Players & players )
+{
+    auto it = mapPlayers.find( mapFileName );
+    if ( it != mapPlayers.end() ) {
```
Got it.

ihhub commented:
Hi @vasilenkoalexey this solution still needs some correction. If you start a multiplayer game and then start a standard game you can observe that the number of human players is more than 1 which is not possible for a standard game. The same stick could be applied even for multiplayer game when we have the number of human players not equal to the selection by user.

vasilenkoalexey commented:
> Hi @vasilenkoalexey this solution still needs some correction. If you start a multiplayer game and then start a standard game you can observe that the number of human players is more than 1 which is not possible for a standard game. The same stick could be applied even for multiplayer game when we have the number of human players not equal to the selection by user.

As far as I understand, you already fixed this in the last commit... Or should we compare the game type in `LoadPlayers`?

ihhub commented:
> > Hi @vasilenkoalexey this solution still needs some correction. If you start a multiplayer game and then start a standard game you can observe that the number of human players is more than 1 which is not possible for a standard game. The same stick could be applied even for multiplayer game when we have the number of human players not equal to the selection by user.
> 
> As far as I understand, you already fixed this in the last commit... Or should we compare the game type in `LoadPlayers`?

Hi @vasilenkoalexey sadly I didn't. I just added an extra check in case of similar map names. We should compare game types as well to avoid the issue I described before.

vasilenkoalexey commented:
> Hi @vasilenkoalexey sadly I didn't. I just added an extra check in case of similar map names. We should compare game types as well to avoid the issue I described before.

Hello @ihhub, I added a comparison of the number of human players. This solved the problem.

ihhub commented:
@vasilenkoalexey a bunch of thanks for fixing this issue!
